### PR TITLE
Tolerate missing v2s2

### PR DESCRIPTION
**Pull request by midnightercz**
For single arch repositories which are not for amd64 manifest
v2sch2 can missing and only manifest list with v2ch1 present.
Method for fetching missing digests should tolerate this.

### CODE REVIEW

File: pubtools/_quay/push_docker.py [hunk 1]
```diff
@@ -617,12 +617,21 @@ def fetch_missing_push_items_digests(self, push_items, target_settings):
                 v2_sch2_cache = {}
                 for tag in tags:
                     for mtype in missing_media_types:
```
Comment (querti) [inline pubtools/_quay/push_docker.py:619]: Can you explain this method to me? If V2S2 manifest wasn't pushed, it tries to get V2S2 manifest from ??somewhere?? But it's also possible that V2S2 manifest is not there? When would such a scenario occur?

**Review by querti — CHANGES_REQUESTED** [review]

File: pubtools/_quay/push_docker.py [hunk 2]
```diff
@@ -617,12 +617,21 @@ def fetch_missing_push_items_digests(self, push_items, target_settings):
...
-                                v2_sch2_cache[repo] = self._fetch_digest(internal_repo, tag, mtype)
-                            item.metadata["new_digests"].setdefault((repo, tag), {})[
-                                mtype
-                            ] = v2_sch2_cache[repo]
+                                try:
+                                    v2_sch2_cache[repo] = self._fetch_digest(
+                                        internal_repo, tag, mtype
+                                    )
+                                # Tolarate missing v2ch2 manifest in the case repo is
+                                # no amd64 arch and has only manifest list and sch2v1
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/push_docker.py:628]: Yeah, that true. ML can contains only v2ch2 manifests. My point here is that repo doesn't have any tag to v2ch2 manifest. I'll update the comment

File: pubtools/_quay/push_docker.py [hunk 3]
```diff
@@ -617,12 +617,21 @@ def fetch_missing_push_items_digests(self, push_items, target_settings):
...
+                                except ManifestNotFoundError:
+                                    pass
+                                else:
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/push_docker.py:631]: else case is executed when there's no exception raised

File: pubtools/_quay/quay_client.py [hunk 4]
```diff
@@ -120,7 +120,13 @@ def get_manifest_digest(self, image, media_type=None):
         Returns (str):
             Manifest digest of the image.
         """
-        manifest = self.get_manifest(image, raw=True, media_type=media_type)
+        try:
+            manifest = self.get_manifest(image, raw=True, media_type=media_type)
+        except requests.exceptions.HTTPError as exc:
+            if exc.response.status_code == 404:
+                raise ManifestNotFoundError()
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/quay_client.py:127]: ManifestTypeError is described as:
Occurs when an incorrect manifest type is encountered
However ManifestNotFoundError is raised when registry returns 404 on manifest

File: pubtools/_quay/push_docker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (midnightercz) [reply pubtools/_quay/push_docker.py:631]: Actually that won't work either, I need to change this

File: pubtools/_quay/push_docker.py (hunk 1 above)
Comment (midnightercz) [reply pubtools/_quay/push_docker.py:619]: v2ch2 can be created on fly when requested by user. But only if there's v2ch1 and not list. If there's list, quay takes list as superior format and pull everything from there. So if there's list but amd64 architecture is not in list, you'll get 404

File: tests/test_quay_client.py [hunk 5]
```diff
@@ -634,3 +636,29 @@ def test_get_manifest_digest():
         assert m.call_count == 1
 
         assert digest == "sha256:b9742c91f353022604e8ed4cf4ab1d688114fc4b133f0e11cbf7dd6272753ac8"
+
+
+def test_get_manifest_digest_not_found():
```
[the remark this answers is not in the document]
Comment (midnightercz) [reply tests/test_quay_client.py:641]: tag can exists only for manifest list so it's real world scenario

**Review by emilyzheng — APPROVED** [review]

**Review by querti — APPROVED** [review]

**Pull request merged**